metadata/src/main/java/org/apache/kafka/controller/ControllerMetricsManager.java

```diff
@@ -0,0 +1,316 @@
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.metadata.BrokerRegistrationChangeRecord;
+import org.apache.kafka.common.metadata.FenceBrokerRecord;
+import org.apache.kafka.common.metadata.MetadataRecordType;
+import org.apache.kafka.common.metadata.PartitionChangeRecord;
+import org.apache.kafka.common.metadata.PartitionRecord;
+import org.apache.kafka.common.metadata.RegisterBrokerRecord;
+import org.apache.kafka.common.metadata.RemoveTopicRecord;
+import org.apache.kafka.common.metadata.TopicRecord;
+import org.apache.kafka.common.metadata.UnfenceBrokerRecord;
+import org.apache.kafka.common.metadata.UnregisterBrokerRecord;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.metadata.BrokerRegistrationFencingChange;
+import org.apache.kafka.metadata.placement.PartitionAssignment;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import static org.apache.kafka.metadata.LeaderConstants.NO_LEADER;
+import static org.apache.kafka.metadata.LeaderConstants.NO_LEADER_CHANGE;
+
+/**
+ * Type for updating controller metrics based on metadata records.
+ */
+final class ControllerMetricsManager {
+    private final static class PartitionState {
+        final int leader;
+        final PartitionAssignment assignment;
+
+        PartitionState(int leader, PartitionAssignment assignment) {
+            this.leader = leader;
+            this.assignment = assignment;
+        }
+
+        int leader() {
+            return leader;
+        }
+
+        PartitionAssignment assignment() {
+            return assignment;
+        }
+    }
+
+    private final Set<Integer> registeredBrokers = new HashSet<>();
+
+    private final Set<Integer> fencedBrokers = new HashSet<>();
+
+    private int topicCount = 0;
+
+    private final Map<TopicIdPartition, PartitionState> topicPartitions = new HashMap<>();
+
+    private final Set<TopicIdPartition> offlineTopicPartitions = new HashSet<>();
+
+    private final Set<TopicIdPartition> imbalancedTopicPartitions = new HashSet<>();
+
+    private final ControllerMetrics controllerMetrics;
+
+    ControllerMetricsManager(ControllerMetrics controllerMetrics) {
+        this.controllerMetrics = controllerMetrics;
+    }
+
+    void replayBatch(long baseOffset, List<ApiMessageAndVersion> messages) {
+        int i = 1;
+        for (ApiMessageAndVersion message : messages) {
+            try {
+                replay(message.message());
+            } catch (Exception e) {
+                String failureMessage = String.format(
+                    "Unable to update controller metrics for %s record, it was %d of %d record(s) " +
+                    "in the batch with baseOffset %d.",
+                    message.message().getClass().getSimpleName(),
+                    i,
+                    messages.size(),
+                    baseOffset
+                );
+                throw new IllegalArgumentException(failureMessage, e);
```

Comment:
Just wondering what our options are here.  Seems to me that ignoring the error applying the metric is one.  Given that we applied the record successfully but fail applying the change in metrics, might we prioritize availability over consistency here?  By throwing an exception we seem to have chosen consistency over availability, and I'm not sure that is the right choice.  WDYT?

Reply:
For metadata, we have agreed that consistency is more important than availability. Inconsistent metadata in the controllers has an impact on the entire cluster.

Reply:
> For metadata, we have agreed that consistency is more important than availability. Inconsistent metadata in the controllers has an impact on the entire cluster.

Agree, but if the records apply, and the metrics do not, we simply have inconsistent metrics but consistent metadata.  We have inconsistent metrics now, without this patch.  So I'm wondering if maybe I am missing something?  If not, I'm still wondering if we should really fail the batch here.

Reply:
That is not true in general. In theory, it is possible for `ControllerManagerTest` to throw and for the other manager to also be in an inconsistent state. In those cases, we want the the controller's error metric to be non-zero.

For cluster monitoring it is really important for the controller to report consistent metrics and to not silently continue running with incorrect metrics.